plugins/Toolbox/src/Toolbox.py

```diff
@@ -0,0 +1,515 @@
+        self.metadataChanged.emit()
+        self._restart_required = True
+        self.restartRequiredChanged.emit()
+
+    @pyqtSlot(str)
+    def enable(self, plugin_id: str):
+        self._plugin_registry.enablePlugin(plugin_id)
+        self.enabledChanged.emit()
+        Logger.log("i", "%s was set as 'active'.", plugin_id)
+        self._restart_required = True
+        self.restartRequiredChanged.emit()
+
+    @pyqtSlot(str)
+    def disable(self, plugin_id: str):
+        self._plugin_registry.disablePlugin(plugin_id)
+        self.enabledChanged.emit()
+        Logger.log("i", "%s was set as 'deactive'.", plugin_id)
+        self._restart_required = True
+        self.restartRequiredChanged.emit()
+
+    @pyqtProperty(bool, notify = metadataChanged)
+    def dataReady(self):
+        return self._packages_model is not None
+
+    @pyqtProperty(bool, notify = restartRequiredChanged)
+    def restartRequired(self):
+        return self._restart_required
+
+    @pyqtSlot()
+    def restart(self):
+        self._package_manager._removeAllScheduledPackages()
+        CuraApplication.getInstance().windowClosed()
+
+
+
+    # Checks
+    # --------------------------------------------------------------------------
+    @pyqtSlot(str, result = bool)
+    def canUpdate(self, package_id: str) -> bool:
+        local_package = self._package_manager.getInstalledPackageInfo(package_id)
+        if local_package is None:
+            return False
+
+        remote_package = None
+        for package in self._metadata["packages"]:
+            if package["package_id"] == package_id:
+                remote_package = package
+        if remote_package is None:
+            return False
+
+        local_version = local_package["package_version"]
+        remote_version = remote_package["package_version"]
+        return Version(remote_version) > Version(local_version)
+
+    @pyqtSlot(str, result = bool)
+    def isInstalled(self, package_id: str) -> bool:
+        return self._package_manager.isPackageInstalled(package_id)
+
+    @pyqtSlot(str, result = bool)
+    def isEnabled(self, package_id: str) -> bool:
+        if package_id in self._plugin_registry.getActivePlugins():
+            return True
+        return False
+
+    def loadingComplete(self) -> bool:
+        populated = 0
+        for list in self._metadata.items():
+            if len(list) > 0:
+                populated += 1
+        if populated == len(self._metadata.items()):
+            return True
+        return False
+
+
+
+    # Make API Calls
+    # --------------------------------------------------------------------------
+    def _makeRequestByType(self, type: str):
+        Logger.log("i", "Toolbox: Requesting %s metadata from server.", type)
+        request = QNetworkRequest(self._request_urls[type])
+        request.setRawHeader(*self._request_header)
+        self._network_manager.get(request)
+
+    @pyqtSlot(str)
+    def startDownload(self, url: str):
```

Comment:
This function needs to cancel any ongoing downloads, or fail or block if there are any ongoing downloads.

Reply:
Don't think it really does as it can't be called in the first place. Since it doesn't induce any bugs I won't be changing that for now.